apps/web/src/components/EditorHeader/index.tsx

```diff
@@ -156,106 +103,101 @@ export default function EditorHeader({
     defaultValue: true,
   })
 
+  const [provider, setProvider] = useState<string | undefined>()
+  const [model, setModel] = useState<string | undefined | null>()
+
+  const promptMetadata = useMemo<PromptMetadata | undefined>(() => {
+    if (!metadata?.config) return undefined
+    return {
+      provider: metadata.config.provider as PromptMetadata['provider'],
+      model: metadata.config.model as PromptMetadata['model'],
+    }
+  }, [metadata?.config])
+
   const providersByName = useMemo(() => {
     return providerApiKeys.reduce((acc, data) => {
       acc[data.name] = data
       return acc
     }, {} as IProviderByName)
   }, [isLoading, providerApiKeys])
 
-  const [selectedProvider, setProvider] = useState<string | undefined>()
-  const [selectedModel, setModel] = useState<string | undefined | null>(() =>
-    selectModel({
-      promptMetadata,
-      providersByName,
-    }),
-  )
-
   const providerOptions = useMemo(() => {
     return providerApiKeys.map((apiKey) => ({
       label: apiKey.name,
       value: apiKey.name,
     }))
   }, [providerApiKeys])
-  const isDefaultProvider =
-    selectedProvider === envClient.NEXT_PUBLIC_DEFAULT_PROJECT_ID
-  const modelOptions = useModelsOptions({
-    provider: selectedProvider
-      ? providersByName[selectedProvider]?.provider
-      : undefined,
-    isDefaultProvider,
+  const modelOptions = useModelOptions({
+    provider: provider ? providersByName[provider]?.provider : undefined,
+    name: provider ? providersByName[provider]?.name : undefined,
   })
 
+  // onPromptMetadataChange
   useEffect(() => {
-    const foundProvider = providersByName[promptMetadata?.providerName]
-    if (foundProvider?.name === selectedProvider) return
-
-    setProvider(foundProvider?.name)
-    setModel(undefined)
-  }, [selectedProvider, providersByName, promptMetadata?.providerName])
-
-  useEffect(() => {
-    const model = selectModel({
-      promptMetadata,
-      providersByName,
-    })
+    if (!promptMetadata) return
 
-    if (selectedModel === model) return
+    if (promptMetadata.provider !== provider) {
+      setProvider(promptMetadata.provider)
+    }
 
-    setModel(model)
-  }, [
-    providersByName,
-    selectedModel,
-    promptMetadata?.providerName,
-    promptMetadata?.model,
-  ])
+    if (!promptMetadata.model || promptMetadata.model !== model) {
+      setModel(promptMetadata.model ?? null)
+    }
+  }, [promptMetadata])
 
   const onSelectProvider = useCallback(
-    (value: string) => {
-      const provider = providersByName[value]!
-      if (!provider) return
+    (selectedProvider: string) => {
+      if (!selectedProvider) return
+      if (selectedProvider === provider) return
+
+      let firstModel
+      if (providersByName[selectedProvider]) {
+        firstModel = findFirstModelForProvider({
+          provider: providersByName[selectedProvider],
+          latitudeProvider: envClient.NEXT_PUBLIC_DEFAULT_PROJECT_ID,
+        })
+      }
```

Comment:
I have seen this same if check + firtModel find 3 times. Maybe you could pass the if inside the function.

Reply:
There are 4 different components (4 editors) with similar functionality.... do you mean you want `findFirstModelForProvider` to accept an undefined provider and return no model if so? For me that it's a little strange but will do.

Reply:
I mean, is what you're doing outside no? you init `let model` as undefined and then assign it if the it exists

Reply:
Yes, but I am not calling the function with no provider because.... why call that function if you don't have one? For me it's strange that function may accept no provider.... but don't worry I will change it